src/ImageSharp/Processing/Processors/Filters/ReduceLuminanceProcessor.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.ParallelUtils;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Filters
+{
+    /// <summary>
+    /// Reduces the luminance of the input image, proportially to the luminance of each pixel.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    internal class ReduceLuminanceProcessor<TPixel> : ImageProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SixLabors.ImageSharp.Processing.Processors.Filters.ReduceLuminanceProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="amount">The strength of the luminance reduction filter.</param>
+        public ReduceLuminanceProcessor(float amount) => this.Amount = amount;
+
+        /// <summary>
+        /// Gets the strength factor for the luminance reduction processor.
+        /// </summary>
+        public float Amount { get; }
+
+        /// <inheritdoc/>
+        protected override void OnFrameApply(ImageFrame<TPixel> source, Rectangle sourceRectangle, Configuration configuration)
+        {
+            var interest = Rectangle.Intersect(sourceRectangle, source.Bounds());
+            int startX = interest.X;
+            float amount = this.Amount;
+
+            ParallelHelper.IterateRowsWithTempBuffer<Vector4>(
+                interest,
+                configuration,
+                (rows, vectorBuffer) =>
+                    {
+                        for (int y = rows.Min; y < rows.Max; y++)
+                        {
+                            Span<Vector4> vectorSpan = vectorBuffer.Span;
+                            int length = vectorSpan.Length;
+                            Span<TPixel> rowSpan = source.GetPixelRowSpan(y).Slice(startX, length);
+                            PixelOperations<TPixel>.Instance.ToVector4(configuration, rowSpan, vectorSpan);
+
+                            ref Vector4 baseRef = ref MemoryMarshal.GetReference(vectorSpan);
+
+                            for (int i = 0; i < vectorSpan.Length; i++)
+                            {
+                                ref Vector4 v = ref Unsafe.Add(ref baseRef, i);
+                                Vector4Utils.Premultiply(ref v);
```

Comment:
I'm not sure I see why though 🤔
I mean, that method is just calling this same overload inside a loop, and here I already have that loop in place, as I'm iterating over the pixels in the current row.
Why would that be faster, if not just almost identical in performance (honest question)?

Reply:
Small piece of code executed on a compact list of data structures leads to surprisingly higher processing speed on modern CPU-s because of better cache locality, and better utilization of instruction pipelines.

If you haven't practiced this before, give it a try! This simple thing will lead to miracles in the performance of your code! Good question where to start though, probably google on stuff about CPU caches, and this thing: "Data-oriented design".

Reply:
Oh you're right, I was just thinking about the overall instructions and not about cache locality.
I actually did study that in a few courses during the computer engineering degree, I guess I've just been working exclusively on UWP for too long now so I lost practice and forgot to think about that ahahah
Thanks for the advice, I'll go ahead and change this then! 😄